rust/iso-8601/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+use chrono::{DateTime, FixedOffset, TimeZone};
+
+pub fn parse_datetime(str: &str) -> Option<DateTime<FixedOffset>> {
+    let values = DateTimeValues::from(str);
+    FixedOffset::east_opt(values.timezone.secs()).and_then(|tz| {
+        tz.with_ymd_and_hms(
+            values.year,
+            values.month,
+            values.date,
+            values.hours,
+            values.minutes,
+            values.seconds,
+        )
+        .single()
+    })
+}
+
+struct DateTimeValues {
+    year: i32,
+    month: u32,
+    date: u32,
+    hours: u32,
+    minutes: u32,
+    seconds: u32,
+    timezone: TimezoneValues,
+}
+
+impl DateTimeValues {
+    fn default() -> Self {
+        Self {
+            year: 1970,
+            month: 1,
+            date: 1,
+            hours: 0,
+            minutes: 0,
+            seconds: 0,
+            timezone: TimezoneValues::default(),
+        }
+    }
+
+    fn from(str: &str) -> Self {
+        let mut values = Self::default();
+        values.parse(str, &ParserState::Year);
+        return values;
+    }
+
+    fn parse(&mut self, str: &str, state: &ParserState) {
+        match state {
+            ParserState::Timezone => self.timezone.parse(str, TimezoneParserState::Sign),
+            _ => self.parse_state(str, state),
+        }
+    }
+
+    fn parse_state(&mut self, str: &str, state: &ParserState) {
+        if let Some(from) = state.parse_from(str) {
+            let size = state.size();
+            let field = str
+                .get(from..from + size)
+                .and_then(|value| state.parse(value));
+
+            if let Some(field) = field {
+                if let Some(next) = state.next() {
+                    self.parse(&str[from + size..], &next);
+                }
+                self.set(field);
+            }
+        }
+    }
+
+    fn set(&mut self, field: DateTimeField) {
+        match field {
+            DateTimeField::Year(value) => self.year = value,
+            DateTimeField::Month(value) => self.month = value,
+            DateTimeField::Date(value) => self.date = value,
+            DateTimeField::Hours(value) => self.hours = value,
+            DateTimeField::Minutes(value) => self.minutes = value,
+            DateTimeField::Seconds(value) => self.seconds = value,
+            DateTimeField::Timezone(value) => self.timezone = value,
+        }
+    }
+}
+
+enum DateTimeField {
+    Year(i32),
+    Month(u32),
+    Date(u32),
+    Hours(u32),
+    Minutes(u32),
+    Seconds(u32),
+    Timezone(TimezoneValues),
+}
+
+#[derive(Debug, PartialEq)]
+enum ParserState {
+    Year,
+    Month,
+    Date,
+    Hours,
+    Minutes,
+    Seconds,
+    Timezone,
+}
+
+impl ParserState {
+    fn next(&self) -> Option<Self> {
```

Comment:
This could be done as iterator and rust nerds will force you to do so but I think it's fine.